openedx/core/djangoapps/user_api/learner_skill_levels/api.py

```diff
@@ -0,0 +1,120 @@
+                else:
+                    learner_skill_score.update({skill: score})
+    return learner_skill_score
+
+
+def get_learner_skill_levels(user, top_categories, user_score_mapping, request_user=False):
+    """
+    Evaluates learner's skill levels in the given job category. Only considers skills for the categories
+    and not their sub-categories.
+
+    Params:
+        user (CourseKey): The identifier for the course.
+        top_categories (List, string): A list of fields (as strings) of job categories and their skills.
+        user_score_mapping (Dict, List): A dictionary of lists to save every user's score against each category.
+        request_user (Boolean): A flag to check if the 'user' is request user or not.
+    Returns:
+        None
+    """
+
+    # get a skill to score mapping for every course user has passed
+    learner_skill_score = _generate_skill_score_mapping(user)
+
+    for skill_category in top_categories:
+        user_score = 0.0
+        category_skills = skill_category['skills']
+        if category_skills:
+            # returns the skills list with scores appended
+            skills_with_score = _get_category_skill_score(category_skills, learner_skill_score)
+
+            # generate a dict with only those skills user has learned, will be needed to calculate user score.
+            category_skills_scores = {
+                item['name']: item['score']
+                for item in skills_with_score
+                if item['score'] is not None
+            }
+            user_score = _calculate_user_skill_score(category_skills_scores)
+
+            if request_user:
+                # for request user, we need to return skills with their scores
+                skill_category['skills'] = skills_with_score
+
+                # for request user we need to iterate over sub-categories skills as well and append scores
+                sub_categories = skill_category['skills_subcategories']
+                if sub_categories:
+                    for sub_category in sub_categories:
+                        subcategory_skills = sub_category['skills']
+                        if subcategory_skills:
+                            sub_category['skills'] = _get_category_skill_score(subcategory_skills, learner_skill_score)
+
+        # if user is other than request user, append their score for current category to
+        # user-score mapping to calculate average later on.
+        # If not, append user and average user score.
+        if not request_user:
+            user_score_mapping[skill_category['name']].append(user_score)
+        else:
+            skill_category['user_score'] = user_score
+            sum_score = sum(user_score_mapping[skill_category['name']], 0.0)
+            average_score = sum_score/len(user_score_mapping[skill_category['name']])
+            skill_category['edx_average_score'] = average_score
+
```

Comment:
this whole if-else [line 110-119] should not be part of this method as this makes this method complex. it should be either in the view where `user_score_mapping` is defined or there should be another utility that should be called from the view.

Reply:
I'm more in the favor of keeping view code simple. Also, we don't have current category information available at view level.